sdk/servicebus/azure-servicebus/azure/servicebus/_common/message.py

```diff
@@ -536,9 +482,50 @@ def abandon(self):
         :raises: ~azure.servicebus.common.errors.MessageSettleFailed if message settle operation fails.
         """
         self._is_live('abandon')
-        self._receiver._settle_deferred('abandoned', [self.lock_token])  # pylint: disable=protected-access
+        try:
+            self._receiver._settle_message(SETTLEMENT_ABANDON, [self.lock_token])  # pylint: disable=protected-access
+        except Exception as e:
+            raise MessageSettleFailed("abandon", e)
         self._settled = True
 
     def defer(self):
-        """A DeferredMessage cannot be deferred. Raises `ValueError`."""
-        raise ValueError("Message is already deferred.")
+        """Defer the message.
+
+        This message will remain in the queue but must be received
+        specifically by its sequence number in order to be processed.
+
+        :raises: ~azure.servicebus.common.errors.MessageAlreadySettled if the message has been settled.
+        :raises: ~azure.servicebus.common.errors.MessageLockExpired if message lock has already expired.
+        :raises: ~azure.servicebus.common.errors.SessionLockExpired if session lock has already expired.
+        :raises: ~azure.servicebus.common.errors.MessageSettleFailed if message settle operation fails.
+        """
+        self._is_live('defer')
+        try:
+            self._receiver._settle_message(SETTLEMENT_DEFER, [self.lock_token])  # pylint: disable=protected-access
+        except Exception as e:
+            raise MessageSettleFailed("defer", e)
+        self._settled = True
+
+    def renew_lock(self):
+        """Renew the message lock.
+
+        This will maintain the lock on the message to ensure
+        it is not returned to the queue to be reprocessed. In order to complete (or otherwise settle)
+        the message, the lock must be maintained. Messages received via ReceiveAndDelete mode are not
+        locked, and therefore cannot be renewed. This operation can also be performed as a threaded
+        background task by registering the message with an `azure.servicebus.AutoLockRenew` instance.
+        This operation is only available for non-sessionful messages.
+
+        :raises: TypeError if the message is sessionful.
+        :raises: ~azure.servicebus.common.errors.MessageLockExpired is message lock has already expired.
+        :raises: ~azure.servicebus.common.errors.MessageAlreadySettled is message has already been settled.
+        """
+        if self._receiver._session_id:  # pylint: disable=protected-access
+            raise TypeError("Session messages cannot be renewed. Please renew the Session lock instead.")
+        self._is_live('renew')
```

Comment:
Given how we're using _is_live, it's probably better to rename it to e.g. "_ensure_is_live" to communicate the imperitive semantics and lack of return value (_is_* implies a boolean, typically)

Reply:
In future we should investigate what the experience is if we remove this pre-check and use a try/raise approach.

Reply:
move the topic to here: https://github.com/Azure/azure-sdk-for-python/issues/10602